octopusdeploy/resource_certificate.go

```diff
@@ -0,0 +1,199 @@
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("error reading certificate %s: %s", certificateId, err.Error())
+	}
+
+	d.Set("name", certificate.Name)
+	d.Set("notes", certificate.Notes)
+	d.Set("environment_ids", certificate.EnvironmentIds)
+	d.Set("tenanted_deployment_participation", certificate.TenantedDeploymentParticipation)
+	d.Set("tenant_ids", certificate.TenantIds)
+	d.Set("tenant_tags", certificate.TenantTags)
+
+	return nil
+}
+
+func buildCertificateResource(d *schema.ResourceData) *octopusdeploy.Certificate {
+	certificateName := d.Get("name").(string)
+
+	var notes string
+	var certificateData string
+	var password string
+	var environmentIds []string
+	var tenantedDeploymentParticipation string
+	var tenantIds []string
+	var tenantTags []string
+
+	notesInterface, ok := d.GetOk("notes")
+	if ok {
+		notes = notesInterface.(string)
+	}
+
+	certificateDataInterface, ok := d.GetOk("certificate_data")
+	if ok {
+		certificateData = certificateDataInterface.(string)
+	}
+
+	passwordInterface, ok := d.GetOk("password")
+	if ok {
+		password = passwordInterface.(string)
+	}
+
+	environmentIdsInterface, ok := d.GetOk("environment_ids")
+	if ok {
+		environmentIds = getSliceFromTerraformTypeList(environmentIdsInterface)
+	}
+
+	if environmentIds == nil {
+		environmentIds = []string{}
+	}
+
+	tenantedDeploymentParticipationInterface, ok := d.GetOk("tenanted_deployment_participation")
+	if ok {
+		tenantedDeploymentParticipation = tenantedDeploymentParticipationInterface.(string)
+	}
+
+	tenantIdsInterface, ok := d.GetOk("tenant_ids")
+	if ok {
+		tenantIds = getSliceFromTerraformTypeList(tenantIdsInterface)
+	}
+
+	if tenantIds == nil {
+		tenantIds = []string{}
+	}
+
+	tenantTagsInterface, ok := d.GetOk("tenant_tags")
+	if ok {
+		tenantTags = getSliceFromTerraformTypeList(tenantTagsInterface)
+	}
+
+	if tenantTags == nil {
+		tenantTags = []string{}
+	}
+
+	var certificate = octopusdeploy.NewCertificate(certificateName, octopusdeploy.SensitiveValue{NewValue: certificateData}, octopusdeploy.SensitiveValue{NewValue: password})
+	certificate.Notes = notes
+	certificate.EnvironmentIds = environmentIds
+	certificate.TenantedDeploymentParticipation = tenantedDeploymentParticipation
+	certificate.TenantIds = tenantIds
+	certificate.TenantTags = tenantTags
+
+	return certificate
+}
+
+func resourceCertificateCreate(d *schema.ResourceData, m interface{}) error {
+	client := m.(*octopusdeploy.Client)
+
+	newCertificate := buildCertificateResource(d)
+	certificate, err := client.Certificate.Add(newCertificate)
+
+	if err != nil {
+		return fmt.Errorf("error creating certificate %s: %s", newCertificate.Name, err.Error())
+	}
+
+	d.SetId(certificate.ID)
+
+	return nil
+}
+
+func resourceCertificateUpdate(d *schema.ResourceData, m interface{}) error {
+	certificate := buildCertificateResource(d)
+	certificate.ID = d.Id()
+
+	client := m.(*octopusdeploy.Client)
+
+	updatedCertificate, err := client.Certificate.Update(certificate)
```

Comment:
I think this should use the `replace` function instead of `update` if the certificate itself has changed (based on thumbprint).

Reply:
Not sure I follow. There is no "Replace" method on the Certificate currently. Update uses this API method: https://demo.octopus.com/swaggerui/index.html#/Certificates/ModifyResponseDescriptor_Certificate_CertificateResource_ Are you saying that API method should not be used?

Reply:
I think the consensus was that for Terraform Update is the right operation (see https://octopusdeploy.slack.com/archives/CDTCSG469/p1561332416096400)

Reply:
I don't have access to that Slack link.

I don't think always replacing the certificate on any change is the right move. Only certain fields should force a new resource. I've marked the `certificate_data` and `password` fields with `ForceNew: true` so that if those fields change Terraform will force a delete / create cycle, but for other fields it will simply update. Does that meet the requirements for this?

Reply:
Ah interesting, it's the conversation in our shared channel on 23/24th Jun. The `replace` function keeps the same ID, but copies the previous certificate to a new ID and archives it. This means that a record is kept of the old certificate without needing to change the ID everywhere it's used. The `update` function just replaces the values of the certificate without a copy/archive.

Relevant parts of the conversation from @MJRichardson our resident expert on that Octopus feature:
> I’m not sure we should ever `replace` via terraform.
> I think it would only make sense if we were specifying that it was a replace, and the cert being replaced was explicitly specified.
>I can’t imagine why you would want to do that via terraform
> It’s designed so that when a cert expires, you can replace it without having to update the projects.
>But the old cert gets kept, and put in the archive.  Just in case.  Because we wanted to encourage people to use :octopusdeploy: as a cert store, being able to retrieve the old certificate seemed like a good idea
> Replacing a cert feels like an imperative action by nature; not something that seems natural to terraform.  Maybe?

Reply:
So I don't think the latest change fits. If anything the `ForceNew` should be doing a `replace` and not a `delete`/`create`.